### PR TITLE
PSR-17 support

### PR DESCRIPTION
**Pull request by boesing**
|    Q          |   A
|-------------- | ------
| BC Break      | no
| New Feature   | yes

### Description

As already done in mezzio/mezzio#75, mezzio/mezzio-router#23 and mezzio/mezzio-problem-details#19 this PR provides support for PSR-17 response factory.

I started removing some prophecy stuff but this is getting too much thus I stopped. Guess I've removed about 50% but I want to focus more on the PSR-17 implementation now. Sorry for not finishing it.

### CODE REVIEW

**Comment by Ocramius** [comment]
Removing milestone: I don't have the understanding (right now) to review this patch, so I requested a review from @weierophinney.

Today, I'll focus on PHP 8.1 support.

**Comment by tobias-trozowski** [comment]
is there any greater benefit for making seperate internal methods instead of assertAttributeSame (which uses reflection)?
IMO adding methods (even if they are marked as `@internal`) solely for the sake of unit testing should really be avoided. 

bumping php to 7.4 might also be suitable as 7.3 reached EOL.

**Comment by boesing** [comment]
https://github.com/sebastianbergmann/phpunit/issues/3338

**Comment by tobias-trozowski** [comment]
in case of the `MezzioUrlGeneratorFactory` the non-failure of the factory is the proof that the instance is constructed accordingly. the optional `?ServerUrlHelper $serverUrlHelper` of `MezzioUrlGenerator` could be validated by calling the `generate` method as it would return different values based on the null|non-null value of `private $serverUrlHelper`.

the same principle would apply to `$urlHelperServiceName` of `MezzioUrlGeneratorFactory`, by providing different values the result is different.

`getMap` can be replaced quite easily by just using `MetadataMap::get(string $class)`.

it is a bit a roundabout way to do it but at least it would not litter up the public interface with test-only things.

**Comment by boesing** [comment]
Since I stopped working on this as of now, I'd be happy to hand this over.
Definitely have time for assistance but no time to work on this myself, so if you do have time and want to work on this, that would be very appreciated. Happy to see your changes in a new PR. 
Is that something you are willing to contribute?

**Comment by tobias-trozowski** [comment]
@boesing i created a new PR ["PSR-17 support #46"](https://github.com/mezzio/mezzio-hal/pull/46) based on your work

File: src/Psr17ResponseFactoryTrait.php [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    private function createResponseFactoryFromDeprecatedCallable(
+        ContainerInterface $container
+    ): ResponseFactoryInterface {
+        /** @var callable():ResponseInterface $responseFactory */
+        $responseFactory = $container->get(ResponseInterface::class);
+
+        return new CallableResponseFactoryDecorator($responseFactory);
+    }
+
+    private function doesConfigurationProvidesDedicatedResponseFactory(ContainerInterface $container): bool
```
Comment (weierophinney) [inline src/Psr17ResponseFactoryTrait.php:46]:
```suggestion
    private function configurationProvidesDedicatedResponseFactory(ContainerInterface $container): bool
```

Reads with a bit more flow.

File: src/Psr17ResponseFactoryTrait.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        $dependencies = $config['dependencies'] ?? [];
+        Assert::isMap($dependencies);
+
+        $delegators = $dependencies['delegators'] ?? [];
+        $aliases    = $dependencies['aliases'] ?? [];
+        Assert::isArrayAccessible($delegators);
+        Assert::isArrayAccessible($aliases);
+
+        if (isset($delegators[ResponseInterface::class]) || isset($aliases[ResponseInterface::class])) {
+            // Even tho, aliases could point to a different service, we assume that there is a dedicated factory
```
Comment (weierophinney) [inline src/Psr17ResponseFactoryTrait.php:63]:
```suggestion
            // Even though aliases could point to a different service, we assume that there is a dedicated factory
```

**Review by weierophinney — APPROVED** [review]
@boesing I rebased against 2.3.x as we've already released 2.2.0. When I did, I bumped the minimum supported PHP to 7.4... and Psalm then flagged a few additional errors, which I fixed, but which were not really related to this.

I've marked a couple nit-picky changes; but otherwise, this makes sense, and I think it's ready to ship.

**Comment by weierophinney** [comment]
Closing in favor of #46.

**Pull request closed without merging**